File: api/poker.py
from enum import IntEnum
from typing import List, Optional, Tuple, Dict
import random
from pydantic import BaseModel

class Suit(str):
    HEARTS = 'hearts'
    DIAMONDS = 'diamonds'
    CLUBS = 'clubs'
    SPADES = 'spades'

class HandRank(IntEnum):
    HIGH_CARD = 0
    PAIR = 1
    TWO_PAIR = 2
    THREE_OF_A_KIND = 3
    STRAIGHT = 4
    FLUSH = 5
    FULL_HOUSE = 6
    FOUR_OF_A_KIND = 7
    STRAIGHT_FLUSH = 8
    ROYAL_FLUSH = 9

class Card(BaseModel):
    suit: str
    rank: int # 2-14

    def __str__(self):
        ranks = {14: 'A', 13: 'K', 12: 'Q', 11: 'J'}
        r = ranks.get(self.rank, str(self.rank))
        return f"{r}-{self.suit}"

class HandEvaluation(BaseModel):
    rank: HandRank
    score: int
    name: str
    cards: List[Card]

SUITS = [Suit.HEARTS, Suit.DIAMONDS, Suit.CLUBS, Suit.SPADES]
RANKS = [14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2]

def create_deck() -> List[Card]:
    deck = [Card(suit=s, rank=r) for s in SUITS for r in RANKS]
    random.shuffle(deck)
    return deck

def get_score(rank: HandRank, cards: List[Card]) -> int:
    score = rank.value * 1000000
    power = 4
    for card in cards:
        score += card.rank * (16 ** power)
        power -= 1
    return score

def evaluate_hand(hole_cards: List[Card], community_cards: List[Card]) -> HandEvaluation:
    all_cards = hole_cards + community_cards
    if not all_cards:
        return HandEvaluation(rank=HandRank.HIGH_CARD, score=0, name="Empty", cards=[])

    all_cards.sort(key=lambda c: c.rank, reverse=True)

    # Flush
    suit_counts: Dict[str, List[Card]] = {}
    for c in all_cards:
        if c.suit not in suit_counts:
            suit_counts[c.suit] = []
        suit_counts[c.suit].append(c)
    
    flush_suit = None
    flush_cards = []
    for s, cards in suit_counts.items():
        if len(cards) >= 5:
            flush_suit = s
            flush_cards = cards
            break
    
    # Straight
    unique_rank_cards = []
    seen_ranks = set()
    for c in all_cards:
        if c.rank not in seen_ranks:
            seen_ranks.add(c.rank)
            unique_rank_cards.append(c)
            
    straight_cards = get_straight_cards(unique_rank_cards)

    # Straight Flush
    straight_flush_cards = None
    if flush_suit:
        # Get unique ranks for flush cards
        unique_flush_cards = []
        seen_flush_ranks = set()
        for c in flush_cards:
            if c.rank not in seen_flush_ranks:
                seen_flush_ranks.add(c.rank)
                unique_flush_cards.append(c)
        straight_flush_cards = get_straight_cards(unique_flush_cards)

    if straight_flush_cards:
        rank = HandRank.ROYAL_FLUSH if straight_flush_cards[0].rank == 14 else HandRank.STRAIGHT_FLUSH
        return HandEvaluation(
            rank=rank,
            score=get_score(rank, straight_flush_cards),
            name="Royal Flush" if rank == HandRank.ROYAL_FLUSH else "Straight Flush",
            cards=straight_flush_cards
        )

    # Four of a Kind
    quads = get_n_of_a_kind(all_cards, 4)
    if quads:
        return HandEvaluation(
            rank=HandRank.FOUR_OF_A_KIND,
            score=get_score(HandRank.FOUR_OF_A_KIND, quads),
            name="Four of a Kind",
            cards=quads
        )

    # Full House
    full_house = get_full_house(all_cards)
    if full_house:
        return HandEvaluation(
            rank=HandRank.FULL_HOUSE,
            score=get_score(HandRank.FULL_HOUSE, full_house),
            name="Full House",
            cards=full_house
        )

    # Flush
    if flush_cards:
        best_flush = flush_cards[:5]
        return HandEvaluation(
            rank=HandRank.FLUSH,
            score=get_score(HandRank.FLUSH, best_flush),
            name="Flush",
            cards=best_flush
        )

    # Straight
    if straight_cards:
        return HandEvaluation(
            rank=HandRank.STRAIGHT,
            score=get_score(HandRank.STRAIGHT, straight_cards),
            name="Straight",
            cards=straight_cards
        )

    # Three of a Kind
    trips = get_n_of_a_kind(all_cards, 3)
    if trips:
        return HandEvaluation(
            rank=HandRank.THREE_OF_A_KIND,
            score=get_score(HandRank.THREE_OF_A_KIND, trips),
            name="Three of a Kind",
            cards=trips
        )

    # Two Pair
    two_pair = get_two_pair(all_cards)
    if two_pair:
        return HandEvaluation(
            rank=HandRank.TWO_PAIR,
            score=get_score(HandRank.TWO_PAIR, two_pair),
            name="Two Pair",
            cards=two_pair
        )

    # Pair
    pair = get_n_of_a_kind(all_cards, 2)
    if pair:
        return HandEvaluation(
            rank=HandRank.PAIR,
            score=get_score(HandRank.PAIR, pair),
            name="Pair",
            cards=pair
        )

    # High Card
    high_card = all_cards[:5]
    return HandEvaluation(
        rank=HandRank.HIGH_CARD,
        score=get_score(HandRank.HIGH_CARD, high_card),
        name="High Card",
        cards=high_card
    )

def get_straight_cards(unique_rank_cards: List[Card]) -> Optional[List[Card]]:
    if len(unique_rank_cards) < 5:
        return None
        
    # Check normal straights
    for i in range(len(unique_rank_cards) - 4):
        chunk = unique_rank_cards[i:i+5]
        if chunk[0].rank - chunk[4].rank == 4:
            return chunk
            
    # Check A-5-4-3-2
    ranks = [c.rank for c in unique_rank_cards]
    if 14 in ranks and 5 in ranks and 4 in ranks and 3 in ranks and 2 in ranks:
        # Construct 5,4,3,2,A
        ace = next(c for c in unique_rank_cards if c.rank == 14)
        others = [c for c in unique_rank_cards if c.rank in [5,4,3,2]]
        # others is sorted desc: 5,4,3,2
        return others + [ace]
        
    return None

def get_n_of_a_kind(cards: List[Card], n: int) -> Optional[List[Card]]:
    counts = {}
    for c in cards:
        if c.rank not in counts:
            counts[c.rank] = []
        counts[c.rank].append(c)
    
    for rank in RANKS:
        if rank in counts and len(counts[rank]) == n:
            main_cards = counts[rank]
            kickers = [c for c in cards if c.rank != rank][:5-n]
            return main_cards + kickers
            
    return None

def get_full_house(cards: List[Card]) -> Optional[List[Card]]:
    counts = {}
    for c in cards:
        if c.rank not in counts:
            counts[c.rank] = []
        counts[c.rank].append(c)
        
    three = None
    two = None
    
    for rank in RANKS:
        if rank in counts and len(counts[rank]) >= 3:
            three = counts[rank][:3]
            break
            
    if three:
        for rank in RANKS:
            if rank in counts and len(counts[rank]) >= 2 and rank != three[0].rank:
                two = counts[rank][:2]
                break
                
    if three and two:
        return three + two
    return None

def get_two_pair(cards: List[Card]) -> Optional[List[Card]]:
    counts = {}
    for c in cards:
        if c.rank not in counts:
            counts[c.rank] = []
        counts[c.rank].append(c)
        
    pairs = []
    for rank in RANKS:
        if rank in counts and len(counts[rank]) >= 2:
            pairs.append(counts[rank][:2])
            
    if len(pairs) >= 2:
        best_pairs = pairs[0] + pairs[1]
        remaining = [c for c in cards if c.rank != pairs[0][0].rank and c.rank != pairs[1][0].rank]
        return best_pairs + [remaining[0]]
        
    return None

File: src/utils/poker.ts
import { Card, HandEvaluation, HandRank, Rank, Suit } from '../types/poker';

export const SUITS: Suit[] = ['hearts', 'diamonds', 'clubs', 'spades'];
export const RANKS: Rank[] = [14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2];

export const createDeck = (): Card[] => {
  const deck: Card[] = [];
  for (const suit of SUITS) {
    for (const rank of RANKS) {
      deck.push({ suit, rank });
    }
  }
  return shuffleDeck(deck);
};

export const shuffleDeck = (deck: Card[]): Card[] => {
  const newDeck = [...deck];
  for (let i = newDeck.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [newDeck[i], newDeck[j]] = [newDeck[j], newDeck[i]];
  }
  return newDeck;
};

export const formatRank = (rank: Rank): string => {
  switch (rank) {
    case 14: return 'A';
    case 13: return 'K';
    case 12: return 'Q';
    case 11: return 'J';
    default: return rank.toString();
  }
};

export const evaluateHand = (holeCards: Card[], communityCards: Card[]): HandEvaluation => {
  const allCards = [...holeCards, ...communityCards];
  if (allCards.length === 0) {
    return { rank: HandRank.HighCard, score: 0, name: 'Empty', cards: [] };
  }
  
  // Sort by rank descending
  allCards.sort((a, b) => b.rank - a.rank);

  // Check Flush
  const flushSuit = getFlushSuit(allCards);
  const flushCards = flushSuit ? allCards.filter(c => c.suit === flushSuit) : [];

  // Check Straight
  const straightCards = getStraightCards(allCards);

  // Check Straight Flush
  let straightFlushCards: Card[] | null = null;
  if (flushSuit && flushCards.length >= 5) {
     straightFlushCards = getStraightCards(flushCards);
  }

  if (straightFlushCards) {
    return {
      rank: straightFlushCards[0].rank === 14 ? HandRank.RoyalFlush : HandRank.StraightFlush,
      score: getScore(HandRank.StraightFlush, straightFlushCards),
      name: straightFlushCards[0].rank === 14 ? 'Royal Flush' : 'Straight Flush',
      cards: straightFlushCards
    };
  }

  // Four of a Kind
  const fourKind = getNOfAKind(allCards, 4);
  if (fourKind) {
    return {
      rank: HandRank.FourOfAKind,
      score: getScore(HandRank.FourOfAKind, fourKind),
      name: 'Four of a Kind',
      cards: fourKind
    };
  }

  // Full House
  const fullHouse = getFullHouse(allCards);
  if (fullHouse) {
    return {
      rank: HandRank.FullHouse,
      score: getScore(HandRank.FullHouse, fullHouse),
      name: 'Full House',
      cards: fullHouse
    };
  }

  // Flush
  if (flushCards.length >= 5) {
    const bestFlush = flushCards.slice(0, 5);
    return {
      rank: HandRank.Flush,
      score: getScore(HandRank.Flush, bestFlush),
      name: 'Flush',
      cards: bestFlush
    };
  }

  // Straight
  if (straightCards) {
    return {
      rank: HandRank.Straight,
      score: getScore(HandRank.Straight, straightCards),
      name: 'Straight',
      cards: straightCards
    };
  }

  // Three of a Kind
  const threeKind = getNOfAKind(allCards, 3);
  if (threeKind) {
    return {
      rank: HandRank.ThreeOfAKind,
      score: getScore(HandRank.ThreeOfAKind, threeKind),
      name: 'Three of a Kind',
      cards: threeKind
    };
  }

  // Two Pair
  const twoPair = getTwoPair(allCards);
  if (twoPair) {
    return {
      rank: HandRank.TwoPair,
      score: getScore(HandRank.TwoPair, twoPair),
      name: 'Two Pair',
      cards: twoPair
    };
  }

  // Pair
  const pair = getNOfAKind(allCards, 2);
  if (pair) {
    return {
      rank: HandRank.Pair,
      score: getScore(HandRank.Pair, pair),
      name: 'Pair',
      cards: pair
    };
  }

  // High Card
  const highCard = allCards.slice(0, 5);
  return {
    rank: HandRank.HighCard,
    score: getScore(HandRank.HighCard, highCard),
    name: 'High Card',
    cards: highCard
  };
};

// Helper functions

const getFlushSuit = (cards: Card[]): Suit | null => {
  const counts: Record<string, number> = {};
  for (const card of cards) {
    counts[card.suit] = (counts[card.suit] || 0) + 1;
    if (counts[card.suit] >= 5) return card.suit;
  }
  return null;
};

const getStraightCards = (cards: Card[]): Card[] | null => {
  const uniqueRankCards: Card[] = [];
  const seenRanks = new Set<number>();
  
  for (const card of cards) {
    if (!seenRanks.has(card.rank)) {
      seenRanks.add(card.rank);
      uniqueRankCards.push(card);
    }
  }

  // Check for 5-high straight (A-5-4-3-2)
  if (seenRanks.has(14) && seenRanks.has(2) && seenRanks.has(3) && seenRanks.has(4) && seenRanks.has(5)) {
    // Check if we have A, 5, 4, 3, 2 consecutive in uniqueRankCards (ignoring gaps)
    // Actually simpler: construct the straight
    const ace = uniqueRankCards.find(c => c.rank === 14)!;
    const five = uniqueRankCards.find(c => c.rank === 5)!;
    const four = uniqueRankCards.find(c => c.rank === 4)!;
    const three = uniqueRankCards.find(c => c.rank === 3)!;
    const two = uniqueRankCards.find(c => c.rank === 2)!;
    
    // Check normal straights first, as they are higher
    // But wait, A-2-3-4-5 is the lowest straight, so checking normal first covers higher ones.
  }

  for (let i = 0; i <= uniqueRankCards.length - 5; i++) {
    const chunk = uniqueRankCards.slice(i, i + 5);
    if (chunk[0].rank - chunk[4].rank === 4) {
      return chunk;
    }
  }

  // Special check for A-5-4-3-2
  if (seenRanks.has(14) && seenRanks.has(5) && seenRanks.has(4) && seenRanks.has(3) && seenRanks.has(2)) {
    const ace = uniqueRankCards.find(c => c.rank === 14)!;
    const others = uniqueRankCards.filter(c => [5, 4, 3, 2].includes(c.rank));
    return [ ...others, ace ]; // 5,4,3,2,A (technically 5 high straight)
  }

  return null;
};

const getNOfAKind = (cards: Card[], n: number): Card[] | null => {
  const counts: Record<number, Card[]> = {};
  for (const card of cards) {
    if (!counts[card.rank]) counts[card.rank] = [];
    counts[card.rank].push(card);
  }

  for (const rank of RANKS) { // Check from high to low
    if (counts[rank] && counts[rank].length === n) {
      const mainCards = counts[rank];
      const kickers = cards.filter(c => c.rank !== rank).slice(0, 5 - n);
      return [...mainCards, ...kickers];
    }
  }
  return null;
};

const getFullHouse = (cards: Card[]): Card[] | null => {
  const counts: Record<number, Card[]> = {};
  for (const card of cards) {
    if (!counts[card.rank]) counts[card.rank] = [];
    counts[card.rank].push(card);
  }

  let three: Card[] | null = null;
  let two: Card[] | null = null;

  for (const rank of RANKS) {
    if (counts[rank] && counts[rank].length >= 3) {
      three = counts[rank].slice(0, 3);
      break;
    }
  }

  if (three) {
    for (const rank of RANKS) {
      if (counts[rank] && counts[rank].length >= 2 && rank !== three[0].rank) {
        two = counts[rank].slice(0, 2);
        break;
      }
    }
  }

  if (three && two) {
    return [...three, ...two];
  }
  return null;
};

const getTwoPair = (cards: Card[]): Card[] | null => {
  const counts: Record<number, Card[]> = {};
  for (const card of cards) {
    if (!counts[card.rank]) counts[card.rank] = [];
    counts[card.rank].push(card);
  }

  const pairs: Card[][] = [];
  for (const rank of RANKS) {
    if (counts[rank] && counts[rank].length >= 2) {
      pairs.push(counts[rank].slice(0, 2));
    }
  }

  if (pairs.length >= 2) {
    const bestTwoPairs = [...pairs[0], ...pairs[1]];
    const remaining = cards.filter(c => c.rank !== pairs[0][0].rank && c.rank !== pairs[1][0].rank);
    return [...bestTwoPairs, remaining[0]];
  }
  return null;
};

const getScore = (rank: HandRank, cards: Card[]): number => {
  // Score = Rank * 1,000,000 + c1*16^4 + c2*16^3 + ...
  // Using base 16 (hex) for card ranks (2-14) fits nicely
  let score = rank * 1000000;
  let power = 4;
  for (const card of cards) {
    score += card.rank * Math.pow(16, power);
    power--;
  }
  return score;
};
